src/oneD/IonFlow.cpp

```diff
@@ -0,0 +1,416 @@
+{
+    // Reference:
+    // Pederson, Timothy, and R. C. Brown. 
+    // "Simulation of electric field effects in premixed methane flames, Comb.
+    // " Flames 94 (1993): 433-448.
+    for (size_t j = j0; j < j1; j++) {
+        double wtm = m_wtm[j];
+        double rho = density(j);
+        double dz = z(j+1) - z(j);
+
+        // mixture-average diffusion
+        double sum = 0.0;
+        for (size_t k = 0; k < m_nsp; k++) {       
+            m_flux(k,j) = m_wt[k]*(rho*m_diff[k+m_nsp*j]/wtm);
+            m_flux(k,j) *= (X(x,k,j) - X(x,k,j+1))/dz;
+            sum -= m_flux(k,j); 
+        }
+
+        // ambipolar diffusion
+        double E_ambi = E(x,j);
+        for (size_t k : m_kCharge) {
+            double Yav = 0.5 * (Y(x,k,j) + Y(x,k,j+1));
+            double drift = rho * Yav * E_ambi
+                           * m_speciesCharge[k] * m_mobi[k+m_nsp*j]; 
+            m_flux(k,j) += drift;
+        }
+
+        // correction flux
+        double sum_flux = 0.0;
+        for (size_t k = 0; k < m_nsp; k++) {
+            sum_flux -= m_flux(k,j); // total net flux
+        }
+        double sum_ion = 0.0;
+        for (size_t k : m_kCharge) {
+            sum_ion += Y(x,k,j);
+        }
+        // The portion of correction for ions is taken off
+        for (size_t k : m_kNeutral) {       
+            m_flux(k,j) += Y(x,k,j) / (1-sum_ion) * sum_flux;
+        }
+    }   
+}
+
+void IonFlow::setSolvingPhase(const size_t phase)
+{
+    if (phase == 1 || 2 || 3) {
+        m_phase = phase;
+    } else {
+        throw CanteraError("IonFlow::updateDiffFluxes",
+                    "solution phase must be set to:"
+                    "1: generate a good guess for the charged species."
+                    "2: Use Prager's ambipolar diffusion model to"
+                    "   calculate the electrical drift."
+                    "3: Use Poisson's equation to solve the E-field,"
+                    "   and replace the drift term in Prager's model"
+                    "   with the new drift term calculated from the E-field");
+    }
+}
+
+void IonFlow::setElectricVoltage(const double v1, const double v2)
+{
+    // This method can be used when you want to add external voltage
+    m_inletVoltage = v1;
+    m_outletVoltage = v2;
+}
+
+void IonFlow::eval(size_t jg, doublereal* xg,
+                  doublereal* rg, integer* diagg, doublereal rdt)
+{
+    StFlow::eval(jg, xg, rg, diagg, rdt);
+    // start of local part of global arrays
+    doublereal* x = xg + loc();
+    doublereal* rsd = rg + loc();
+    integer* diag = diagg + loc();
+
+    size_t jmin, jmax;
+    if (jg == npos) { // evaluate all points
+        jmin = 0;
+        jmax = m_points - 1;
+    } else { // evaluate points for Jacobian
+        size_t jpt = (jg == 0) ? 0 : jg - firstPoint();
+        jmin = std::max<size_t>(jpt, 1) - 1;
+        jmax = std::min(jpt+1,m_points-1);
+    }
+
+    if (m_phase == 3) {
```

Comment:
You could simplify this by just writing
```c++
if (m_phase != 3) {
    return
}
```
after the call to the parent class method.